### ISSUE TITLE
Rename fmu.iteration to fmu.ensemble

### ISSUE THREAD

**Issue by perolavsvendsen** [equinor/fmu-dataio#459]
Large question, and requires some discussion and then some planning. But as it keeps coming up, I think we need to discuss it:

Basic question is, should we rename the current `fmu.iteration` to `fmu.ensemble`?

Possible reasons could include:
- More in line with terminology used in e.g. ERT, Webviz, daily speech, etc.
- All iterations are ensembles (?), but not all ensembles are iterations

**Comment by jcrivenaes** [comment]
From Chat GPT:

> In the context of scientific research and forecasting, an "ensemble" refers to a technique that combines multiple predictive models to obtain better accuracy and reliability than could be achieved from any single model. The ensemble method leverages the diversity of multiple models to improve predictions, manage uncertainties, and reduce the risk of significant errors. Here are some key points that define an ensemble in this context
> ....
> Overall, an ensemble approach represents a powerful tool in data science and predictive modeling, allowing for more reliable and accurate predictions by harnessing the collective insights of multiple models.

Notice **multi-model** is possible

Since `iteration` in our context is restricted to one model setup, I would strongly advice against renaming iterations to ensembles

**Comment by perolavsvendsen** [comment]
There is a point to be made on internal alignment inside FMU as well. Not saying that anyone has the "right" definitions here (we are all in our own little bubbles), but I note this from https://ert.readthedocs.io/en/latest/developer_documentation/concepts.html#ensemble:

> ## Ensemble
> As the name hints towards, ERT’s approach to studying a template model, is to create an ensemble of realisations. In particular, an ensemble is a list of realisations. The idea is that with a sufficiently large ensemble the uncertainty of the template model is represented by the ensemble.

**Comment by anders-kiaer** [comment]
Webviz uses the word `ensemble` (similar to ERT) where `fmu-dataio` uses `iteration`.

> All iterations are ensembles (?), but not all ensembles are iterations

This summarizes also my mental picture of how we use the terminology in Equinor. Iterations I feel are usually referring to the AHM updating process in ERT. Using `iteration` for prediction feels a bit unnatural IMO.

> In the context of scientific research and forecasting, an "ensemble" refers to a technique that combines multiple predictive models to obtain better accuracy and reliability than could be achieved from any single model. The ensemble method leverages the diversity of multiple models to improve predictions, manage uncertainties, and reduce the risk of significant errors. Here are some key points that define an ensemble in this context

I interpret ChatGPT's answer here actually as being consistent with ERT's usage of the word `ensemble`. Instead ChatGPT (and statistical physics, which it looks like it bases its answer on) differ instead on how the term `model` is used as I see it. What we call `realization` in FMU, ChatGPT call `model`.

ChatGPT's answer if we replace `model` with `realization`:

> In the context of scientific research and forecasting, an "ensemble" refers to a technique that combines multiple predictive realizations to obtain better accuracy and reliability than could be achieved from any single realization. The ensemble method leverages the diversity of multiple realizations to improve predictions, manage uncertainties, and reduce the risk of significant errors. Here are some key points that define an ensemble in this context
....
Overall, an ensemble approach represents a powerful tool in data science and predictive modeling, allowing for more reliable and accurate predictions by harnessing the collective insights of multiple realizations.

**Comment by anders-kiaer** [comment]
This is ERT's definition of ensemble ( https://ert.readthedocs.io/en/latest/developer_documentation/concepts.html#ensemble )

> **Ensemble**
> As the name hints towards, ERT’s approach to studying a template model, is to create an ensemble of realisations. In particular, an ensemble is a list of realisations. The idea is that with a sufficiently large ensemble the uncertainty of the template model is represented by the ensemble.

which is aligned with Wikipedia's definition ( https://en.wikipedia.org/wiki/Ensemble_%28mathematical_physics%29 ), where "**virtual copies of a system**" corresponds to "**realizations of the subsurface system/model**" in FMU terminology:

> In [physics](https://en.wikipedia.org/wiki/Physics), specifically [statistical mechanics](https://en.wikipedia.org/wiki/Statistical_mechanics), an ensemble (also statistical ensemble) is an idealization consisting of a large number of virtual copies (sometimes infinitely many) of a [system](https://en.wikipedia.org/wiki/System), considered all at once, each of which represents a possible state that the real system might be in.

**Comment by perolavsvendsen** [comment]
"Sister"-issue on Sumo: https://github.com/equinor/sumo/issues/193

**Comment by equinor-ruaj** [comment]
We're assuming this will be implemented as a straight up renaming, i.e. no other changes except for name.
Is that what you intend to do @tnatt ?

**Comment by tnatt** [comment]
Yes, no other changes to the fields in the pydantic model @equinor-ruaj 👍 . 
But in the transition we want both `fmu.iteration` and `fmu.ensemble` to be present right, as duplicates of each other? 

And what about the Iteration and Realization objects that you use in Sumo? Can we go fully over to `fmu.ensemble` (i.e. just rename  `fmu.iteration`) for those or should we keep both of them there also? 
The changes will be added to a new schema version.

**Comment by tnatt** [comment]
We also need to consider what to do about `fmu.context.stage`. If we want this renaming to be consistent, we should transition from `iteration` to `ensemble` also here.

**Comment by equinor-ruaj** [comment]
Yes, we want both in the transition 👍 
I think we want both for Iteration and Realization objects for a period as well.

Do you have any input @rwiker ?

**Comment by rwiker** [comment]
Good point about `fmu.context.stage`; this should be set to `ensemble`for new objects in `fmu-dataio`. For existing objects, we'll eventually have to run a bulk update.

For now, I think I'd like to have both `fmu.iteration.*` and `fmu.ensemble.*` populated by `fmu-dataio`. The `iteration` objects are created by the the uploader; this is something we can handle in the Sudo team.

At some point in the future we will have to bulk-update all objects that have information in `fmu.iteration.*` but nothing in `fmu.ensemble.*`.

Before we start adding documents using `fmu.ensemble` we should update the Elasticsearch mappings. Do we have any other changes than adding `fmu.ensemble.{name,uuid}` in parallel with `fmu.iteration.{name,uuid}`?
